docs/RationaleSeq.md

```diff
@@ -0,0 +1,100 @@
+
+- Physical devices:
+  - **Unclocked Latch:** A memory element which is only sensitive to the
+  levels of its inputs. Has no clock. Example: SR Latch.
+  - Clocked (gated) **latch:** A latch wherein the inputs are gated by a
+  clock. Transparent the entire time the clock is high. Generally referred to
+  as a "latch". Examples: "gated SR latch", "D latch".
+  - Edge-triggered **flip-flop:** An edge-sensitive memory element. Captures
+  the input value on one or both clock edges. Variants: posedge FF, negedge
+  FF, "edge-sensitive" FF (captures the input value on both edges),
+  resettable FF.
+- Abstract models:
+  - **Register:** A synchronous, resettable memory element. Can be
+  implemented using any of the above "circuit level" elements.
+
+## The register operation
+
+The `seq.reg` op models an abstract notion of a "register", independent of
+its implementation (e.g. latch, D flip-flop). It is intended to be 'lowered'
+to a specific implementation, which may be an `sv.always_ff` block, a device
+primitive instantiation, or even a feedback loop in all `comb` logic. Our
+intention is to allow analysis and optimization of sequential logic without
+having to reason about timing-dependent behavior. This makes it somewhat
+distinct from the SystemVerilog register wherein one has to worry about reset
+behavior and implementation (latch vs. flip-flop).
+
+`Reg` has four operands:
+
+- **input**: The value to be captured at 'clock'. Generally called 'q'.
+Accepts any type, results in the same type.
+- **clock**: Capture 'value' on the positive edge of this signal.
+- **reset**: Signal to set the state to 'resetValue'. Optional.
+- **resetValue**: A value which the state is set to upon reset. Required iff
+'reset' is present.
+
+```mlir
+%d = seq.reg %input, %clk [, %reset [, %resetValue] ] : $type(input)
+```
+
+### Rationale
+
+Several design decisions were made in defining this op. Mostly, they were
+made to simplify it while still providing the common case. Providing support
+for all flavors of registers is an anti-goal of this op. If an omitted
+feature is needed, it can be added or (if not common or precludes
+optimization) another op could be added.
+
+- Logical features:
+  - Inclusion of optional 'reset' signal: This operand makes lowering to an
+  efficient implementation of reset easier. Omission of it would require some
+  (potentially complex) analysis to find the reset mux if required (which it
+  usually is to ensure efficient synthesis).
+  - Inclusion of 'resetValue': if we have a 'reset' signal, we need to
+  include a value.
+  - Omission of 'enable': enables are easily modeled via a multiplexer on the
+  input with one of the mux inputs as the output of the register. This
+  property makes 'enables' easier to detect than reset in the common case.
+
+- Timing / clocking:
+  - Omission of 'negedge' event on 'clock': this is easily modeled by
+  inverting the clock value.
```

Comment:
This seems like it might lead to problems, but going with the simpler thing here (as written) seems a good start.